### PR TITLE
[reconfiguration] another attempt to fix the bug and better test cove…

### PR DESCRIPTION
**Pull request by zekun000**
…rage

## Motivation

https://github.com/diem/diem/pull/8879 didn't fix the problem because I used the wrong function and tests didn't cover it.

This PR fixes the problem as well as making sure test covers the edge case. 

The fix is in executed_block.rs.

For test, specifically we forces a timeout (by sending proposals to half of the validators) on the direct child block of reconfiguration when failpoints is enabled, so it'll deterministically reproduce the error.

Also we enabled failpoints in forge local build, so all tests involving multi nodes and epoch change would cover the missing case.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

cargo x test -p smoke-test -- genesis --nocapture

the test failed without the fix, and passed with the fix.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

**Comment by bmwill** [comment]
The PR title indicates that it includes better test coverage...I don't see any tests in this PR, am I missing something?

I wouldn't really consider failpoints sufficient to provide test coverage either.

**Comment by zekun000** [comment]
@bmwill the summary has more details, it adds the logic when failpoints is enabled any smoke test that involves multi nodes and reconfiguration covers the edge case

can you elaborate more on failpoints not sufficient to provide test coverage?

**Comment by bmwill** [comment]
I would instead expect to see a test explicitly testing this regression instead of claiming that failpoints are sufficient to solve the problem. Ideally as a concensus unit test.

**Comment by zekun000** [comment]
it's not something an unit test can fully cover, it's an e2e functionality that spans across consensus/execution/move. An integration test (smoke test) should cover this if we could trigger this code path which is exactly what failpoints does.

the behavior is controlled by the failpoints feature, we could switch it to something else like a config but feature seems the most natural way to cover it in both smoke tests and cluster tests (with fault injection).

another option is to mock everything else but only test the consensus part in an unit test but it wouldn't cover everything.

**Comment by msmouse** [comment]
I'd prioritize pulling this fix in over adding more testing specific to this issue. But I agree with @bmwill it's the best to have a purpose built unit test for this. Even something naive would have value, like `assert_eq!(make_suffix_block_result().transactions_to_commit(), vec![]`

And injecting the error "deterministically" makes me a little nervous tbh -- the happy path becomes the edge case with failpoints enabled.

**Comment by zekun000** [comment]
added a small unit test inspired by @msmouse

File: consensus/src/round_manager.rs [hunk 1]
```diff
@@ -913,4 +930,25 @@ impl RoundManager {
...
+        #[cfg(feature = "failpoints")]
+        {
+            let direct_suffix = block_data.is_reconfiguration_suffix()
+                && !block_data
+                    .quorum_cert()
+                    .parent_block()
+                    .has_reconfiguration();
+            let continuous_round =
+                block_data.round() == block_data.quorum_cert().certified_block().round() + 1;
+            direct_suffix && continuous_round
```
Comment (msmouse) [inline consensus/src/round_manager.rs:949]: How about give this 50% chance of failing?

**Review by msmouse — DISMISSED** [review]

File: consensus/src/round_manager.rs (hunk 1 above)
Comment (zekun000) [reply consensus/src/round_manager.rs:949]: are you worried that happy path is not tested? we have plenty of chances testing that e.g. cluster test runs a 100 epoch tests without failpoints.
